framework/source/class/qx/core/MObjectId.js

```diff
@@ -0,0 +1,231 @@
+        }
+      }
+    },
+    
+    /**
+     * Returns the object with the specified ID
+     * 
+     * @param id
+     *          {String} ID of the object
+     * @return {qx.core.Object?} the found object
+     */
+    getObject: function(id) {
+      if (this.__ownedObjects) {
+        var obj = this.__ownedObjects[id];
+        if (obj !== undefined) {
+          return obj;
+        }
+      }
+      
+      // Handle paths
+      if (id.indexOf('/') > -1) {
+        var segs = id.split('/');
+        var target = this;
+        var found = segs.every(function(seg) {
+          if (!seg.length) {
+            return true;
+          }
+          if (!target) {
+            return false;
+          }
+          var tmp = target.getObject(seg);
+          if (tmp !== undefined) {
+            target = tmp;
+            return true;
+          }
+        });
+        return found ? target : undefined;
+      }
+      
+      // No object, creating the object
+      var obj = this._createObjectImpl(id);
+      if (obj !== undefined) {
+        obj.setObjectId(id);
+        this.addOwnedObject(obj);
+      }
+      
+      return obj;
+    },
+    
+    /**
+     * Creates the object, intended to be overridden. Null is a valid return
+     * value and will be cached by `getObject`, however `undefined` is NOT a
+     * valid value and so will not be cached meaning that `_createObjectImpl`
+     * will be called multiple times until a valid value is returned.
+     * 
+     * @param id {String} ID of the object
+     * @return {qx.core.Object?} the created object
+     */
+    _createObjectImpl: function(id) {
+      return;  // Return undefined
+    },
+    
+    /**
+     * Adds an object as owned by this object
+     * 
+     * @param obj {qx.core.Object} the object to register
+     */
+    addOwnedObject: function(obj) {
+      if (!this.__ownedObjects) {
+        this.__ownedObjects = {};
+      }
+      var id = obj.getObjectId();
+      if (!id) {
+        throw new Error("Cannot register an object that has no ID, obj=" + obj);
+      }
+      if (this.__ownedObjects[id]) {
+        throw new Error("Cannot register an object with ID '" + id + "' because that ID is already in use, this=" + this + ", obj=" + obj);
+      }
+      if (obj.getOwner() != null) {
+        throw new Error("Cannot register an object with ID '" + id + "' because it is already owned by another object this=" + this + ", obj=" + obj);
+      }
+      obj.__changingOwner = true;
+      try {
+        obj.setOwner(this);
+      } finally {
+        obj.__changingOwner = false;
+      }
+      this.__ownedObjects[id] = obj;
+    },
+
+    /**
+     * Discards an object from the list of owned objects; note that this does
+     * not dispose of the object, simply forgets it if it exists.
+     * 
+     * @param id {String|Object} the ID of the object to discard, or the object itself
+     */
+    discardOwnedObject: function(id) {
+      if (!this.__ownedObjects) {
```

Comment:
What do you think about having this stuff done in a private function, e.g., `__discardOwnedObject`, that also has a `force` argument? Then you could have `addOwnedObject` call `__discardOwnedObject` to discard it from a prior owner. The `force` argument, if `true` would ignore `obj.getOwner()` returning null, and allow the `__changingOwner` member to potentially be set to true.

`discardOwnedObject` would call `__discardOwnedObject` with the `foce` value being false.

The goal, if not obvious, is to allow *changing* ownership with a call to `addOwnedObject` just like a widget can be moved from one container to another by simply calling `add()`.

Reply:
I did start coding something like that, but because of the additional complication of making sure that IDs are unique within owners when changing owner and/or ID, the processes involved started to get quite complex.  

As the pattern here is to create objects on demand, so there has to be a public `discard` method to undo the on-demand `_createObjectImpl`; also, conceptually it seems that owned objects will normally remain in one owner anyway, so reverting to a situation where it has to be manually removed before adding to another owner seemed acceptable.

On reflection though, I think a couple of points:
* rename `discardOwnedObject` to `removeOwnedObject` for better consistency with the add/remove
* `removeOwnedObject` to allow silently ignore IDs that don't exist in case they have just not been created yet
* and Ill make `removeOwnedObject` remove it from it's current owner